lib/gc/Analysis/MatmulConfigAnalysis.cpp

```diff
@@ -0,0 +1,443 @@
+                                     const MatmulConfig &config,
+                                     SystemDesc &sysDesc) {
+  auto M = shape[0], N = shape[1], K = shape[2];
+  auto dtypeSize = DataLayout().getTypeSizeInBits(
+      ShapeAdaptor(linalgOp.getDpsInputs()[1].getType()).getElementType());
+  // if use K split, there will be one more final reduce and break the post
+  // fusion
+
+  auto KSplitPenalty = 8.0 * (dtypeSize / bitPerByte);
+  auto memoryConsumptionPerThread =
+      M * K * 1.0 / config.MThreads / config.KThreads +
+      K * N * 1.0 / config.KThreads / config.NThreads +
+      M * N * ((config.KThreads - 1) * KSplitPenalty + 1.0) / config.MThreads /
+          config.NThreads;
+  return memoryConsumptionPerThread;
+}
+
+double computationIntensityOnL2Cache(linalg::LinalgOp &linalgOp,
+                                     ArrayRef<uint32_t> shape,
+                                     const MatmulConfig &config,
+                                     SystemDesc &sysDesc) {
+  double simulationPenalty = 0.7;
+  auto L2Cache = sysDesc.getCacheSize(2);
+  auto dtypeSize = DataLayout().getTypeSizeInBits(
+      ShapeAdaptor(linalgOp.getDpsInputs()[1].getType()).getElementType());
+  auto outOfCachePenalty = 1024;
+  double FLOPS = 2.0 * config.MBlock * config.NBlock * config.KBlock;
+  double memoryConsumption = config.MBlock * config.NBlock +
+                             config.NBlock * config.KBlock +
+                             config.MBlock * config.KBlock;
+  double computationIntensity = FLOPS / memoryConsumption;
+  if (memoryConsumption * (dtypeSize / bitPerByte) >
+      L2Cache * simulationPenalty) {
+    computationIntensity /= outOfCachePenalty;
+  }
+  return 1 / computationIntensity;
+}
+
+using CostModelFn =
+    std::function<float(linalg::LinalgOp &linalgOp, ArrayRef<uint32_t> shape,
+                        MatmulConfig cfg, SystemDesc &sysDesc)>;
+
+std::vector<MatmulConfig>
+filterConfigByCostModel(std::vector<MatmulConfig> configs,
+                        linalg::LinalgOp &linalgOp, ArrayRef<uint32_t> shape,
+                        SystemDesc &sysDesc, CostModelFn costModel,
+                        float eliminationRatio = 0.5, float threshold = -1) {
+  std::vector<MatmulConfig> result;
+  std::vector<float> costs;
+  std::vector<size_t> idx;
+  for (auto [i, config] : llvm::enumerate(configs)) {
+    costs.push_back(costModel(linalgOp, shape, config, sysDesc));
+    idx.push_back(i);
+  }
+  std::stable_sort(idx.begin(), idx.end(), [&costs](size_t i1, size_t i2) {
+    return costs[i1] < costs[i2];
+  });
+  auto thresholdCost = costs[idx[(size_t)(eliminationRatio * configs.size())]];
+  thresholdCost =
+      threshold < thresholdCost && threshold > 0 ? threshold : thresholdCost;
+  for (size_t i = 0; i < configs.size(); i++) {
+    if (costs[idx[i]] <= thresholdCost) {
+      result.push_back(configs[idx[i]]);
+    }
+  }
+  llvm::errs() << "thresholdCost is: " << thresholdCost
```

Comment:
Will you change it to  `LLVM_DEBUG(llvm::dbgs() xxxx)`?

Reply:
Thanks for reminding! I will warp these with `LLVM_DEBUG`